File: src/common/UnionFind.cpp
#include "UnionFind.h"

namespace Penciloid
{
UnionFind::UnionFind() : size(0), parent(nullptr)
{
}

UnionFind::UnionFind(int size) : size(size)
{
	parent = new int[size];
	for (int i = 0; i < size; ++i) parent[i] = -1;
}

UnionFind::~UnionFind()
{
	if (parent) delete[] parent;
}

bool UnionFind::Join(int p, int q)
{
	p = Root(p);
	q = Root(q);

	if (p == q) return false;

	if (parent[p] > parent[q]) {
		parent[q] += parent[p];
		parent[p] = q;
	} else {
		parent[p] += parent[q];
		parent[q] = p;
	}

	return true;
}

}

File: player/slitherlink2/unionfind.js
function UnionFind(len) {
	this.value = new Array;
	for (i = 0; i < len; ++i) this.value.push(-1);
}

UnionFind.prototype.root = function (v) {
	return this.value[v] < 0 ? v : (this.root(this.value[v]));
}

UnionFind.prototype.join = function (u, v) {
	u = this.root(u);
	v = this.root(v);

	if (u == v) return false;

	// this.value[u] += this.value[v];
	this.value[v] = u;
	return true;
}